File: C#/Dapper_ORM/DAL/Repositories/TerritoryRepository.cs
using DAL.Repositories;
using DAL.Entities;
using System.Data.SqlClient;
using Dapper;

namespace DAL.Repositories;

public class TerritoryRepository /*: GenericRepository<Territory, string>*/
{
    private readonly string _connectionString;
    public TerritoryRepository(string connectionString) /*: base(connectionString, "Territories", "TerritoryID")*/
    {
        _connectionString = connectionString;
    }
    [Obsolete]
    public async Task<List<Territory>> GetAllAsync()
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"SELECT * FROM Territories";
            var result = await connection.QueryAsync<Territory>(query);
            return result.ToList();
        }
    }
    [Obsolete]
    public async Task<Territory?> GetByIdAsync(string id)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            connection.Open();
            var query = $"SELECT * FROM Territories WHERE TerritoryID = @Id";
            return await connection.QueryFirstOrDefaultAsync<Territory>(query, new { Id = id });
        }
    }
    [Obsolete]
    public async Task<Territory> AddAsync(Territory entity)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"INSERT INTO Territories (TerritoryID, TerritoryDescription, RegionID) VALUES (@TerritoryID, @TerritoryDescription, @RegionID); SELECT CAST(SCOPE_IDENTITY() as int);";
            var id = await connection.QuerySingleAsync<int>(query, entity);
            var propertyInfo = typeof(Territory).GetProperty("TerritoryID");
            propertyInfo.SetValue(entity, id);
            return entity;
        }
    }
    [Obsolete]
    public async Task UpdateAsync(Territory entity)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"UPDATE Territories SET TerritoryDescription = @TerritoryDescription, RegionID = @RegionID WHERE TerritoryID =  @TerritoryID";
            await connection.ExecuteAsync(query, entity);
        }
    }
    [Obsolete]
    public async Task DeleteAsync(string id)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"DELETE FROM Territories WHERE TerritoryID = @Id";
            await connection.ExecuteAsync(query, new { Id = id });
        }
    }
}

File: Go/GORM_ORM/dal/repositories/territoryRepository.go
package repositories

import (
	"GORM_ORM/dal/entities"

	"gorm.io/gorm"
)

type TerritoryRepository interface {
	GetAll() ([]entities.Territory, error)
	GetByID(id string) (*entities.Territory, error)
	Create(entity *entities.Territory) error
	Update(entity *entities.Territory) error
	Delete(id string) error
}

type territoryRepository struct {
	genericRepo GenericRepository[entities.Territory]
}

func NewTerritoryRepository(db *gorm.DB) TerritoryRepository {
	return &territoryRepository{
		genericRepo: NewGenericRepository[entities.Territory](db),
	}
}

func (r *territoryRepository) GetAll() ([]entities.Territory, error) {
	return r.genericRepo.GetAll()
}

func (r *territoryRepository) GetByID(id string) (*entities.Territory, error) {
	return r.genericRepo.GetByID(id)
}

func (r *territoryRepository) Create(entity *entities.Territory) error {
	return r.genericRepo.Create(entity)
}

func (r *territoryRepository) Update(entity *entities.Territory) error {
	return r.genericRepo.Update(entity)
}

func (r *territoryRepository) Delete(id string) error {
	return r.genericRepo.Delete(id)
}
